File: DotSpatial.Positioning.Compact/GPS/Examples/Desktop Framework/VB.NET/MainForm.vb
Imports DotSpatial.Positioning
Imports DotSpatial.Positioning.Gps
Imports DotSpatial.Positioning.Gps.IO
Imports DotSpatial.Positioning.Gps.Nmea
Imports System.Threading

Public Class MainForm

    ' In case no GPS device can be found, we can use a GPS device emulator.
    ' The VirtualDevice class encapsulates all GPS simulation features.

    Public Sub New()

        ' This call is required by the Windows Form Designer.
        InitializeComponent()

        ' Hook into GPS.NET's device detection events.  These events will report on
        ' any GPS devices which have been found, along with any problems encountered and reasons
        ' why a particular device could NOT be detected.
        AddHandler Devices.DeviceDetectionAttempted, AddressOf Devices_DeviceDetectionAttempted
        AddHandler Devices.DeviceDetectionAttemptFailed, AddressOf Devices_DeviceDetectionAttemptFailed
        AddHandler Devices.DeviceDetectionStarted, AddressOf Devices_DeviceDetectionStarted
        AddHandler Devices.DeviceDetectionCompleted, AddressOf Devices_DeviceDetectionCompleted
        AddHandler Devices.DeviceDetectionCanceled, AddressOf Devices_DeviceDetectionCanceled
        AddHandler Devices.DeviceDetected, AddressOf Devices_DeviceDetected

        ' Hook up event handlers for application-level and AppDomain-level exceptions so 
        ' they can be reported to the user
        AddHandler Application.ThreadException, AddressOf Application_ThreadException
        AddHandler AppDomain.CurrentDomain.UnhandledException, AddressOf CurrentDomain_UnhandledException
    End Sub

#Region "  GPS Device Detection Events  "

    Sub Devices_DeviceDetectionCanceled(ByVal sender As Object, ByVal e As EventArgs)
        BeginInvoke(New EventHandler(AddressOf DeviceDetectionCanceled), sender, e)
    End Sub

    Private Sub DeviceDetectionCanceled(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Device detection canceled!"
        detectButton.Enabled = True
        cancelDetectButton.Enabled = False
    End Sub

    Sub Devices_DeviceDetected(ByVal sender As Object, ByVal e As DeviceEventArgs)
        BeginInvoke(New EventHandler(Of DeviceEventArgs)(AddressOf DeviceDetected), sender, e)
    End Sub

    Private Sub DeviceDetected(ByVal sender As Object, ByVal e As DeviceEventArgs)
        Dim item As ListViewItem
        For Each item In devicesListView.Items
            If Object.ReferenceEquals(item.Tag, e.Device) Then
                item.SubItems(1).Text = "GPS DETECTED"
                item.ImageIndex = 0
            End If
        Next
        devicesListView.Refresh()
    End Sub

    Sub Devices_DeviceDetectionCompleted(ByVal sender As Object, ByVal e As EventArgs)
        BeginInvoke(New EventHandler(AddressOf DeviceDetectionCompleted), sender, e)
    End Sub

    Private Sub DeviceDetectionCompleted(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Device detection complete."
        detectButton.Enabled = True
        cancelDetectButton.Enabled = False
    End Sub

    Sub Devices_DeviceDetectionStarted(ByVal sender As Object, ByVal e As EventArgs)
        BeginInvoke(New EventHandler(AddressOf DeviceDetectionStarted), sender, e)
    End Sub

    Private Sub DeviceDetectionStarted(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Detecting GPS devices..."
        devicesListView.Items.Clear()
        detectButton.Enabled = False
        cancelDetectButton.Enabled = True
    End Sub

    Sub Devices_DeviceDetectionAttemptFailed(ByVal sender As Object, ByVal e As DeviceDetectionExceptionEventArgs)
        BeginInvoke(New EventHandler(Of DeviceDetectionExceptionEventArgs)(AddressOf DeviceDetectionAttemptFailed), sender, e)
    End Sub

    Private Sub DeviceDetectionAttemptFailed(ByVal sender As Object, ByVal e As DeviceDetectionExceptionEventArgs)
        Dim item As ListViewItem
        For Each item In devicesListView.Items
            If Object.ReferenceEquals(item.Tag, e.Device) Then
                item.SubItems(1).Text = e.Exception.Message
                item.ToolTipText = e.Exception.Message
                item.ImageIndex = 1
            End If
        Next item
        devicesListView.Refresh()
    End Sub

    Sub Devices_DeviceDetectionAttempted(ByVal sender As Object, ByVal e As DeviceEventArgs)
        BeginInvoke(New EventHandler(Of DeviceEventArgs)(AddressOf DeviceDetectionAttempted), sender, e)
    End Sub

    Private Sub DeviceDetectionAttempted(ByVal sender As Object, ByVal e As DeviceEventArgs)
        undetectButton.Enabled = True

        For Each existingItem As ListViewItem In devicesListView.Items
            If Object.ReferenceEquals(existingItem.Tag, e.Device) Then
                existingItem.SubItems(1).Text = "Detecting..."
                Return
            End If
        Next

        Dim item As New ListViewItem
        item.Text = e.Device.Name
        item.ImageIndex = 2
        item.Tag = e.Device
        item.SubItems.Add(New ListViewItem.ListViewSubItem(item, "Detecting..."))
        devicesListView.Items.Add(item)
        devicesListView.Refresh()
    End Sub

#End Region

#Region "  NmeaInterpreter Events  "

    ' In GPS.NET 3.0, events may be raised from a thread other than the form's own thread.  As a result,
    ' we must use Invoke or BeginInvoke to "marshal" the code to the Form's thread.

    Private Sub nmeaInterpreter1_SpeedChanged(ByVal sender As Object, ByVal e As SpeedEventArgs) Handles nmeaInterpreter1.SpeedChanged
        BeginInvoke(New EventHandler(Of SpeedEventArgs)(AddressOf SpeedChanged), sender, e)
    End Sub

    Private Sub SpeedChanged(ByVal sender As Object, ByVal e As SpeedEventArgs)
        speedTextBox.Text = e.Speed.ToString()
        speedLabel.Text = speedTextBox.Text
    End Sub

    Private Sub nmeaInterpreter1_BearingChanged(ByVal sender As Object, ByVal e As AzimuthEventArgs) Handles nmeaInterpreter1.BearingChanged
        BeginInvoke(New EventHandler(Of AzimuthEventArgs)(AddressOf BearingChanged), sender, e)
    End Sub

    Private Sub BearingChanged(ByVal sender As Object, ByVal e As AzimuthEventArgs)
        bearingTextBox.Text = e.Azimuth.ToString()
        bearingLabel.Text = bearingTextBox.Text
    End Sub

    Private Sub nmeaInterpreter1_AltitudeChanged(ByVal sender As Object, ByVal e As DistanceEventArgs) Handles nmeaInterpreter1.AltitudeChanged
        BeginInvoke(New EventHandler(Of DistanceEventArgs)(AddressOf AltitudeChanged), sender, e)
    End Sub

    Private Sub AltitudeChanged(ByVal sender As Object, ByVal e As DistanceEventArgs)
        altitudeTextBox.Text = e.Distance.ToString()
        altitudeLabel.Text = altitudeTextBox.Text
    End Sub

    Private Sub nmeaInterpreter1_Paused(ByVal sender As Object, ByVal e As EventArgs) Handles nmeaInterpreter1.Paused
        BeginInvoke(New EventHandler(AddressOf Paused), sender, e)
    End Sub

    Private Sub Paused(ByVal sender As Object, ByVal e As EventArgs)
        pauseButton.Enabled = False
        resumeButton.Enabled = True
        statusLabel.Text = "Paused."
    End Sub

    Private Sub nmeaInterpreter1_Resumed(ByVal sender As Object, ByVal e As EventArgs) Handles nmeaInterpreter1.Resumed
        BeginInvoke(New EventHandler(AddressOf Resumed), sender, e)
    End Sub

    Private Sub Resumed(ByVal sender As Object, ByVal e As EventArgs)
        pauseButton.Enabled = True
        resumeButton.Enabled = False
    End Sub

    Private Sub nmeaInterpreter1_Starting(ByVal sender As Object, ByVal e As DeviceEventArgs) Handles nmeaInterpreter1.Starting
        BeginInvoke(New EventHandler(Of DeviceEventArgs)(AddressOf Starting), sender, e)
    End Sub

    Private Sub Starting(ByVal sender As Object, ByVal e As DeviceEventArgs)
        statusLabel.Text = "Connecting to " + e.Device.Name + "..."
    End Sub

    Private Sub nmeaInterpreter1_Started(ByVal sender As Object, ByVal e As EventArgs) Handles nmeaInterpreter1.Started
        BeginInvoke(New EventHandler(AddressOf Started), sender, e)
    End Sub

    Private Sub Started(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Connected!  Waiting for data..."
        sentenceListBox.Items.Clear()
        startButton.Enabled = False
        stopButton.Enabled = True
        pauseButton.Enabled = True
        resumeButton.Enabled = False

        positionLabel.Text = Position.Empty.ToString()
        speedLabel.Text = Speed.Empty.ToString()
        bearingLabel.Text = Azimuth.Empty.ToString()
        altitudeLabel.Text = Distance.Empty.ToString()
    End Sub

    Private Sub nmeaInterpreter1_Stopping(ByVal sender As Object, ByVal e As EventArgs)
        BeginInvoke(New EventHandler(AddressOf Stopping), sender, e)
    End Sub

    Private Sub Stopping(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Stopping GPS device..."
    End Sub

    Private Sub nmeaInterpreter1_Stopped(ByVal sender As Object, ByVal e As EventArgs)
        BeginInvoke(New EventHandler(AddressOf Stopped), sender, e)
    End Sub

    Private Sub Stopped(ByVal sender As Object, ByVal e As EventArgs)
        statusLabel.Text = "Stopped."
        startButton.Enabled = True
        stopButton.Enabled = False
        pauseButton.Enabled = False
        resumeButton.Enabled = False
    End Sub

    Private Sub nmeaInterpreter1_SatellitesChanged(ByVal sender As Object, ByVal e As SatelliteListEventArgs)
        BeginInvoke(New EventHandler(Of SatelliteListEventArgs)(AddressOf SatellitesChanged), sender, e)
    End Sub

    Private Sub SatellitesChanged(ByVal sender As Object, ByVal e As SatelliteListEventArgs)
        Dim satellite As Satellite

        ' Update each satellite
        For Each satellite In e.Satellites
            Dim isSatelliteNew As Boolean = True

            ' Look through the existing list for matches
            Dim viewItem As ListViewItem
            For Each viewItem In satellitesListView.Items
                ' Each item's Tag property houses a Satellite object
                Dim existing As Satellite = CType(viewItem.Tag, Satellite)

                ' Do the ID's match?
                If existing.PseudorandomNumber.Equals(satellite.PseudorandomNumber) Then
                    ' Yes. Update the listview columns
                    viewItem.SubItems(2).Text = satellite.Azimuth.ToString()
                    viewItem.SubItems(3).Text = satellite.Elevation.ToString()
                    viewItem.SubItems(4).Text = satellite.SignalToNoiseRatio.ToString()
                    isSatelliteNew = False
                End If
            Next

            ' If no existing satellite was found, then add a new one
            If isSatelliteNew Then
                ' Add a new column for this satellite
                Dim newItem As New ListViewItem(satellite.PseudorandomNumber.ToString())
                newItem.SubItems.Add(satellite.Name)
                newItem.SubItems.Add(satellite.Azimuth.ToString())
                newItem.SubItems.Add(satellite.Elevation.ToString())
                newItem.SubItems.Add(satellite.SignalToNoiseRatio.ToString())
                newItem.Tag = satellite
                satellitesListView.Items.Add(newItem)
            End If
        Next
    End Sub

    Private Sub nmeaInterpreter1_PositionChanged(ByVal sender As Object, ByVal e As PositionEventArgs) Handles nmeaInterpreter1.PositionChanged
        BeginInvoke(New EventHandler(Of PositionEventArgs)(AddressOf PositionChanged), sender, e)
    End Sub

    Private Sub PositionChanged(ByVal sender As Object, ByVal e As PositionEventArgs)
        positionTextBox.Text = e.Position.ToString()
        positionLabel.Text = positionTextBox.Text
    End Sub

    Private Sub nmeaInterpreter1_DateTimeChanged(ByVal sender As Object, ByVal e As DateTimeEventArgs) Handles nmeaInterpreter1.DateTimeChanged
        BeginInvoke(New EventHandler(Of DateTimeEventArgs)(AddressOf DateTimeChanged), sender, e)
    End Sub

    Private Sub DateTimeChanged(ByVal sender As Object, ByVal e As DateTimeEventArgs)
        dateTimeTextBox.Text = e.DateTime.ToShortDateString() + " " + e.DateTime.ToLongTimeString()
        utcDateTimeTextBox.Text = e.DateTime.ToUniversalTime().ToString("R")
    End Sub

    Private Sub nmeaInterpreter1_SentenceReceived(ByVal sender As Object, ByVal e As NmeaSentenceEventArgs) Handles nmeaInterpreter1.SentenceReceived
        BeginInvoke(New EventHandler(Of NmeaSentenceEventArgs)(AddressOf SentenceReceived), sender, e)
    End Sub

    Private Sub SentenceReceived(ByVal sender As Object, ByVal e As NmeaSentenceEventArgs)
        If sentenceListBox.Items.Count >= 100 Then
            sentenceListBox.Items.RemoveAt(0)
        End If

        sentenceListBox.Items.Add(e.Sentence.ToString())
        sentenceListBox.SelectedIndex = sentenceListBox.Items.Count - 1

        statusLabel.Text = "Receiving GPS data."
    End Sub

#End Region

#Region "  Button Events  "

    Private Sub detectButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles detectButton.Click
        Devices.BeginDetection()
    End Sub

    Private Sub cancelDetectButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles cancelDetectButton.Click
        Devices.CancelDetection(True)
    End Sub

    Private Sub undetectButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles undetectButton.Click
        Devices.Undetect()
        devicesListView.Items.Clear()
        undetectButton.Enabled = False
    End Sub

    Private Sub startButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles startButton.Click
        Try
            nmeaInterpreter1.Start()
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Cannot connect to GPS")
        End Try
    End Sub

    Private Sub stopButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles stopButton.Click
        nmeaInterpreter1.Stop()
    End Sub

    Private Sub pauseButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles pauseButton.Click
        nmeaInterpreter1.Pause()
    End Sub

    Private Sub resumeButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles resumeButton.Click
        nmeaInterpreter1.Resume()
    End Sub

#End Region

#Region "  Context Menu Events  "

    Private Sub deviceContextMenu_Opening(ByVal sender As System.Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles deviceContextMenu.Opening
        e.Cancel = devicesListView.SelectedItems.Count = 0
    End Sub

    Private Sub redetectMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles redetectMenuItem.Click
        Dim device As Device = DirectCast(devicesListView.SelectedItems(0).Tag, Device)
        device.Undetect()
        device.BeginDetection()
    End Sub

    Private Sub resetMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles resetMenuItem.Click
        Dim device As Device = DirectCast(devicesListView.SelectedItems(0).Tag, Device)
        device.Reset()
    End Sub

#End Region

#Region "  Other Form Control Events  "

    Private Sub devicesListView_ItemSelectionChanged(ByVal sender As Object, ByVal e As ListViewItemSelectionChangedEventArgs) Handles devicesListView.ItemSelectionChanged
        propertyGrid1.SelectedObject = e.Item.Tag
    End Sub

    Private Sub serialCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles serialCheckBox.CheckedChanged
        Devices.AllowSerialConnections = serialCheckBox.Checked
        exhaustiveCheckBox.Enabled = serialCheckBox.Checked
    End Sub

    Private Sub exhaustiveCheckBox_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles exhaustiveCheckBox.CheckedChanged
        Devices.AllowExhaustiveSerialPortScanning = exhaustiveCheckBox.Checked
    End Sub

    Private Sub bluetoothCheckBox_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles bluetoothCheckBox.CheckedChanged
        Devices.AllowBluetoothConnections = bluetoothCheckBox.Checked
    End Sub

    Private Sub firstDeviceCheckBox_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles firstDeviceCheckBox.CheckedChanged
        Devices.IsOnlyFirstDeviceDetected = firstDeviceCheckBox.Checked
    End Sub

    Private Sub clockSynchronizationCheckBox_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles clockSynchronizationCheckBox.CheckedChanged
        Devices.IsClockSynchronizationEnabled = clockSynchronizationCheckBox.Checked
    End Sub

#End Region

#Region "  Unhandled Exception Events  "

	Private Sub CurrentDomain_UnhandledException(ByVal sender As Object, ByVal e As UnhandledExceptionEventArgs)
		Dim ex As Exception = DirectCast(e.ExceptionObject, Exception)
		NotifyOfUnhandledException(ex)
	End Sub

	Private Sub Application_ThreadException(ByVal sender As Object, ByVal e As ThreadExceptionEventArgs)
		NotifyOfUnhandledException(e.Exception)
	End Sub

	''' <summary>
	''' Logs an unhandled exception and displays a message box alerting the user to the error.
	''' </summary>
	''' <param name="exception">The unhandled exception.</param>
	Private Sub NotifyOfUnhandledException(ByVal exception As Exception)
		Try
			' Log the exception (and all of its inner exceptions)
			Dim innerException As Exception = exception
			While (Not (innerException) Is Nothing)
				Trace.TraceError(innerException.ToString)
				innerException = innerException.InnerException
			End While

			' Stop the interpreter
			nmeaInterpreter1.Stop()
		Finally
			' Display the error to the user
			MessageBox.Show( _
				"An unexpected error has occurred." & vbLf & vbLf & exception.GetType.ToString & ": " & exception.Message, _
				Application.ProductName, MessageBoxButtons.OK, MessageBoxIcon.Error)
		End Try
	End Sub

#End Region
End Class

File: Source/Core/DotSpatial.Positioning.Compact/GPS/Examples/Desktop Framework/C#/MainForm.cs
﻿using System;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Windows.Forms;
using DotSpatial.Positioning;
using DotSpatial.Positioning.Gps;
using DotSpatial.Positioning.Gps.Emulators;
using DotSpatial.Positioning.Gps.Nmea;
using DotSpatial.Positioning.Gps.IO;

namespace Diagnostics
{
    public partial class MainForm : Form
    {
        public MainForm()
        {
            InitializeComponent();

            /* Hook into GPS.NET's device detection events.  These events will report on
             * any GPS devices which have been found, along with any problems encountered and reasons
             * why a particular device could NOT be detected.
             */
            Devices.DeviceDetectionAttempted += new EventHandler<DeviceEventArgs>(Devices_DeviceDetectionAttempted);
            Devices.DeviceDetectionAttemptFailed += new EventHandler<DeviceDetectionExceptionEventArgs>(Devices_DeviceDetectionAttemptFailed);
            Devices.DeviceDetectionStarted += new EventHandler(Devices_DeviceDetectionStarted);
            Devices.DeviceDetectionCompleted += new EventHandler(Devices_DeviceDetectionCompleted);
            Devices.DeviceDetectionCanceled += new EventHandler(Devices_DeviceDetectionCanceled);
            Devices.DeviceDetected += new EventHandler<DeviceEventArgs>(Devices_DeviceDetected);

            /* Hook up event handlers for application-level and AppDomain-level exceptions so 
             * they can be reported to the user
             */
            Application.ThreadException += new System.Threading.ThreadExceptionEventHandler(Application_ThreadException);
            AppDomain.CurrentDomain.UnhandledException += new UnhandledExceptionEventHandler(CurrentDomain_UnhandledException);
        }

        #region GPS Device Detection Events

        void Devices_DeviceDetectionCanceled(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Device detection canceled!";
                detectButton.Enabled = true;
                cancelDetectButton.Enabled = false;
            }));
        }

        void Devices_DeviceDetected(object sender, DeviceEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                foreach (ListViewItem item in devicesListView.Items)
                {
                    if (object.ReferenceEquals(item.Tag, e.Device))
                    {
                        item.SubItems[1].Text = "GPS DETECTED";
                        item.ImageIndex = 0;
                    }
                }
                devicesListView.Refresh();
            }));
        }

        void Devices_DeviceDetectionCompleted(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Device detection complete.";
                detectButton.Enabled = true;
                cancelDetectButton.Enabled = false;
            }));
        }

        void Devices_DeviceDetectionStarted(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Detecting GPS devices...";
                devicesListView.Items.Clear();
                detectButton.Enabled = false;
                cancelDetectButton.Enabled = true;
            }));
        }

        void Devices_DeviceDetectionAttemptFailed(object sender, DeviceDetectionExceptionEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                foreach (ListViewItem item in devicesListView.Items)
                {
                    if (object.ReferenceEquals(item.Tag, e.Device))
                    {
                        item.SubItems[1].Text = e.Exception.Message;
                        item.ToolTipText = e.Exception.Message;
                        item.ImageIndex = 1;
                    }
                }
                devicesListView.Refresh();
            }));
        }

        void Devices_DeviceDetectionAttempted(object sender, DeviceEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                undetectButton.Enabled = true;

                foreach (ListViewItem existingItem in devicesListView.Items)
                {
                    if (object.ReferenceEquals(existingItem.Tag, e.Device))
                    {
                        existingItem.SubItems[1].Text = "Detecting...";
                        return;
                    }
                }

                ListViewItem item = new ListViewItem();
                item.Text = e.Device.Name;
                item.ImageIndex = 2;
                item.Tag = e.Device;
                item.SubItems.Add(new ListViewItem.ListViewSubItem(item, "Detecting..."));
                devicesListView.Items.Add(item);
                devicesListView.Refresh();
            }));
        }

        #endregion

        #region NmeaInterpreter Events

        private void nmeaInterpreter1_SpeedChanged(object sender, SpeedEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                speedTextBox.Text = e.Speed.ToString();
                speedLabel.Text = speedTextBox.Text;
            }));

        }

        private void nmeaInterpreter1_BearingChanged(object sender, AzimuthEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                bearingTextBox.Text = e.Azimuth.ToString();
                bearingLabel.Text = bearingTextBox.Text;
            }));
        }

        private void nmeaInterpreter1_AltitudeChanged(object sender, DistanceEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                altitudeTextBox.Text = e.Distance.ToString();
                altitudeLabel.Text = altitudeTextBox.Text;
            }));

        }


        private void nmeaInterpreter1_Paused(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                pauseButton.Enabled = false;
                resumeButton.Enabled = true;
                statusLabel.Text = "Paused.";
            }));
        }

        private void nmeaInterpreter1_Resumed(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                pauseButton.Enabled = true;
                resumeButton.Enabled = false;
            }));
        }

        private void nmeaInterpreter1_Starting(object sender, DeviceEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Connecting to " + e.Device.Name + "...";
            }));
        }

        private void nmeaInterpreter1_Started(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Connected!  Waiting for data...";
                sentenceListBox.Items.Clear();
                startButton.Enabled = false;
                stopButton.Enabled = true;
                pauseButton.Enabled = true;
                resumeButton.Enabled = false;

                positionLabel.Text = Position.Empty.ToString();
                speedLabel.Text = Speed.Empty.ToString();
                bearingLabel.Text = Azimuth.Empty.ToString();
                altitudeLabel.Text = Distance.Empty.ToString();
            }));
        }

        private void nmeaInterpreter1_Stopping(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Stopping GPS device...";
            }));
        }

        private void nmeaInterpreter1_Stopped(object sender, EventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                statusLabel.Text = "Stopped.";
                startButton.Enabled = true;
                stopButton.Enabled = false;
                pauseButton.Enabled = false;
                resumeButton.Enabled = false;
            }));
        }


        private void nmeaInterpreter1_SatellitesChanged(object sender, SatelliteListEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                foreach (Satellite satellite in e.Satellites)
                {
                    bool isSatelliteNew = true;

                    // Look for an existing satellite
                    foreach (ListViewItem viewItem in satellitesListView.Items)
                    {
                        Satellite existing = (Satellite)viewItem.Tag;
                        if (existing.PseudorandomNumber.Equals(satellite.PseudorandomNumber))
                        {
                            // Update shiz
                            viewItem.SubItems[2].Text = satellite.Azimuth.ToString();
                            viewItem.SubItems[3].Text = satellite.Elevation.ToString();
                            viewItem.SubItems[4].Text = satellite.SignalToNoiseRatio.ToString();
                            isSatelliteNew = false;
                        }
                    }

                    // If no existing satellite was found, then add a new one
                    if (isSatelliteNew)
                    {
                        ListViewItem newItem = new ListViewItem(satellite.PseudorandomNumber.ToString());
                        newItem.SubItems.Add(satellite.Name);
                        newItem.SubItems.Add(satellite.Azimuth.ToString());
                        newItem.SubItems.Add(satellite.Elevation.ToString());
                        newItem.SubItems.Add(satellite.SignalToNoiseRatio.ToString());
                        newItem.Tag = satellite;
                        satellitesListView.Items.Add(newItem);
                    }
                }
            }));
        }

        private void nmeaInterpreter1_PositionChanged(object sender, PositionEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                positionTextBox.Text = e.Position.ToString();
                positionLabel.Text = positionTextBox.Text;
            }));
        }

        private void nmeaInterpreter1_DateTimeChanged(object sender, DateTimeEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                dateTimeTextBox.Text = e.DateTime.ToShortDateString() + " " + e.DateTime.ToLongTimeString();
                utcDateTimeTextBox.Text = e.DateTime.ToUniversalTime().ToString("R");
            }));
        }

        private void nmeaInterpreter1_SentenceReceived(object sender, DotSpatial.Positioning.Gps.Nmea.NmeaSentenceEventArgs e)
        {
            BeginInvoke(new MethodInvoker(delegate()
            {
                if (sentenceListBox.Items.Count >= 100)
                sentenceListBox.Items.RemoveAt(0);

                sentenceListBox.Items.Add(e.Sentence.ToString());
                sentenceListBox.SelectedIndex = sentenceListBox.Items.Count - 1;

                statusLabel.Text = "Receiving GPS data.";
            }));
        }

        #endregion

        #region Button Events

        private void detectButton_Click(object sender, EventArgs e)
        {
            Devices.BeginDetection();
        }

        private void cancelDetectButton_Click(object sender, EventArgs e)
        {
            Devices.CancelDetection(true);
        }

        private void undetectButton_Click(object sender, EventArgs e)
        {
            Devices.Undetect();
            devicesListView.Items.Clear();
            undetectButton.Enabled = false;
        }

        private void startButton_Click(object sender, EventArgs e)
        {
            try
            {
                nmeaInterpreter1.Start();
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Cannot connect to GPS");
            }
        }

        private void stopButton_Click(object sender, EventArgs e)
        {
            nmeaInterpreter1.Stop();
        }

        private void pauseButton_Click(object sender, EventArgs e)
        {
            nmeaInterpreter1.Pause();
        }

        private void resumeButton_Click(object sender, EventArgs e)
        {
            nmeaInterpreter1.Resume();
        }

        #endregion

        #region Context Menu Events

        private void deviceContextMenu_Opening(object sender, System.ComponentModel.CancelEventArgs e)
        {
            e.Cancel = devicesListView.SelectedItems.Count == 0;
        }

        private void redetectMenuItem_Click(object sender, EventArgs e)
        {
            Device device = (Device)devicesListView.SelectedItems[0].Tag;
            device.Undetect();
            device.BeginDetection();
        }

        private void resetMenuItem_Click(object sender, EventArgs e)
        {
            Device device = (Device)devicesListView.SelectedItems[0].Tag;
            device.Reset();
        }

        #endregion

        #region Other Form Control Events

        private void devicesListView_ItemSelectionChanged(object sender, ListViewItemSelectionChangedEventArgs e)
        {
            propertyGrid1.SelectedObject = e.Item.Tag;
        }

        private void serialCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            Devices.AllowSerialConnections = serialCheckBox.Checked;
            exhaustiveCheckBox.Enabled = serialCheckBox.Checked;
        }

        private void exhaustiveCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            Devices.AllowExhaustiveSerialPortScanning = exhaustiveCheckBox.Checked;
        }

        private void bluetoothCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            Devices.AllowBluetoothConnections = bluetoothCheckBox.Checked;
        }

        private void firstDeviceCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            Devices.IsOnlyFirstDeviceDetected = firstDeviceCheckBox.Checked;
        }

        private void clockSynchronizationCheckBox_CheckedChanged(object sender, EventArgs e)
        {
            Devices.IsClockSynchronizationEnabled = clockSynchronizationCheckBox.Checked;
        }

        #endregion

        #region Unhandled Exception Events

        void CurrentDomain_UnhandledException(object sender, UnhandledExceptionEventArgs e)
        {
            Exception ex = (Exception)e.ExceptionObject;
            NotifyOfUnhandledException(ex);
        }

        void Application_ThreadException(object sender, ThreadExceptionEventArgs e)
        {
            NotifyOfUnhandledException(e.Exception);
        }

        /// <summary>
        /// Logs an unhandled exception and displays a message box alerting the user to the error.
        /// </summary>
        /// <param name="exception">The unhandled exception.</param>
        private void NotifyOfUnhandledException(Exception exception)
        {
            try
            {
                // Log the exception (and all of its inner exceptions)
                Exception innerException = exception;
                while (innerException != null)
                {
                    Trace.TraceError(innerException.ToString());
                    innerException = innerException.InnerException;
                }

                // Stop the interpreter
                nmeaInterpreter1.Stop();
            }
            finally
            {
                // Display the error to the user
                MessageBox.Show(
                    "An unexpected error has occurred.\n\n" + exception.GetType() + ": " + exception.Message,
                    Application.ProductName,
                    MessageBoxButtons.OK,
                    MessageBoxIcon.Error);
            }
        }

        #endregion
    }
}